docs/generic-interfaces-infrastructure.md

```diff
@@ -0,0 +1,426 @@
+- Projection methods (accounting for semantics differences, if needed)
+- CCW vtable implementation
+- IDIC interface implementation
+- IDIC interface implementation lookup
+
+Supporting all of these scenarios, and doing so as efficiently as possible while also respecting all of CsWinRT 3.0 design goals (eg. proper trimming support), requires several additional types being generated per generic interface instantiation, along with some related APIs exposed from the CsWinRT runtime.
+
+We will use `IEnumerator<T>` (`IIterator<T>`) as reference to document all of these types. There are some differences with other generic interface types, but they all follow the same general principles. This document will also mention such differences where relevant.
+
+> [!NOTE]
+> This document specifically only covers generic **collection** interfaces. Other generic type instantiations, such as delegate types (and associated event source infrastructure, when applicable) and `KeyValuePair<TKey, TValue>`, are handled differently, and are already implemented in `cswinrtgen`. See code there for reference.
+
+## RCW (in `WinRT.Runtime.dll`)
+
+The starting point for RCW scenarios is an RCW class that can be used to marshal interfaces to native. The Windows Runtime interfaces are mapped to different interfaces in the .NET world (eg. `IIterator<T>` -> `IEnumerator<T>`), and sometimes there might be subtle semantic differences, which CsWinRT takes care of hiding. Here is the API surface for the `IEnumerator<T>` RCW:
+
+```csharp
+public abstract class WindowsRuntimeEnumerator<T> : WindowsRuntimeObject, IEnumerator<T>, IWindowsRuntimeInterface<IEnumerator<T>>
+{
+    public WindowsRuntimeEnumerator(WindowsRuntimeObjectReference nativeObjectReference);
+
+    protected sealed override bool IsOverridableInterface(in Guid iid);
+
+    // These will be implemented to invoke the underlying 'IIterator<T>' methods directly
+    protected abstract T CurrentNative { get; }
+    protected abstract bool HasCurrentNative { get; }
+    protected abstract bool MoveNextNative();
+
+    // These will implement the managed 'IEnumerator<T>' interface, with the expected .NET semantics
+    public T Current { get; }
+    object IEnumerator.Current { get; }
+    public bool MoveNext();
+    public void Dispose();
+    public void Reset();
+
+    public WindowsRuntimeObjectReferenceValue GetInterface();
+}
+```
+
+This type derives from `WindowsRuntimeObject`, which provides all the usual support for IDIC, generated COM interfaces, and more. If needed, these RCW types (one per generic interface) can also contain additional state necessary to properly map the Windows Runtime interface behavior to the managed one. This is the case for `IEnumerator<T>`, which is slightly different than `IIterator<T>`, for instance.
+
+Next, an interface is needed for each "Methods" implementation, for each generic interface. For `IIterator<T>`, it looks like this:
```

Comment:
I was mostly with you until this point. "An interface is needed" -- why? How does this connect to the scenarios above? I assume there's some general-purpose C#/WinRT machinery that is trying to reach out for this, but when/where?